### DISCUSSION TITLE
Replacement for add-user-keycloak.sh

### DISCUSSION THREAD

**Discussion by fmarchioni** [keycloak/keycloak#10302] [Archive]
Hi all,
I'm using the following script to add users for testing purpose:
` /opt/jboss/keycloak/bin/add-user-keycloak.sh`
Is there a replacement command to add users from the Command Line?
Thanks
Francesco

**Comment by pedroigor** [comment]
@fmarchioni There isn't. Isn't it enough to import a realm instead?

**Reply by stianst** [reply]
That's a rather complicated alternative, which would require importing a master realm.

**Comment by psytester** [comment]
Same issue here. Running on linux server **without** desktop environment and Xterm forwarding is not possible nor allowed.
How to initally add the admin user if http://localhost:8080/ ist not possible?

BTW, calling the page remotly http://node_ip:8080/ shows me this outdated info.
1. `/auth` path is deprecated.
2. there is no `./bin/add-user-keycloak.sh` script in Quarkus build anymore
``` 
You need local access to create the initial admin user.

Open http://localhost:8080/auth
or use the add-user-keycloak script.
```

**Reply by 19216801-Franz** [reply]
You can add the initial admin user by setting the environment variables before starting the `kc.sh start` command like so:
```
export KEYCLOAK_ADMIN="username"
export KEYCLOAK_ADMIN_PASSWORD="password"
/opt/keycloak/bin/kc.sh start-dev
```

**Reply by psytester** [reply]
Thank you, it works outside of container, too.
Hopefully documentation will be updated.

Now I have another issue because I'm behind a reverse proxy and no TLS setup on keycloak node itself.

**Reply by Wolfsrudel** [reply]
It's already mentioned in the documentation: https://www.keycloak.org/docs/latest/server_admin/index.html#creating-the-account-remotely

**Reply by psytester** [reply]
👍 
:see_no_evil: it was too much to read today.

**Comment by fmarchioni** [comment]
Thanks for your replies to my query.
I was wondering if the older add-user-keycloak.sh could be restored with a
shell script using kcadm.sh

Something like this:

kcadm.sh config credentials --server http://localhost:8080 --realm master
--user $KEYCLOAK_ADMIN --password $KEYCLOAK_ADMIN_PASSWORD
kcadm.sh create users -s username=$1 -s enabled=true -r master
kcadm.sh set-password -r master --username $1 --new-password $2

Overall, it might help to have a script for quickly adding an extra user in
the master realm.


Il giorno mar 22 feb 2022 alle ore 16:15 Wolfsrudel <
***@***.***> ha scritto:

**Comment by vlow** [comment]
The main problem seems to me that `add-user-keycloak.sh` allowed us to add users before Keycloak was started, and starting with Keycloak 17 this is no longer possible.

`add-user-keycloak.sh` made it very easy to add test users for local development by using a custom script as entry point for the Docker Container like that:

```bash
#!/bin/bash
/opt/jboss/keycloak/bin/add-user-keycloak.sh -u test_user_1 -p test_password --realm test_realm \
  --roles role1,role2
/opt/jboss/keycloak/bin/add-user-keycloak.sh -u test_user_2 -p test --realm test_realm \
  --roles role2,role3
/opt/jboss/tools/docker-entrypoint.sh -b 0.0.0.0 -Djboss.bind.address.private=127.0.0.1
```

Since it doesn't seem to be possible to export a realm including the users in the admin UI that would mean I know have to either:

1. start a local instance of Keycloak with my realm
2. create my test users
3. stop Keycloak and - since it's not possible to export a realm including users in the admin ui - export the realm using something like
```
-Dkeycloak.migration.action=export
-Dkeycloak.migration.provider=singleFile
-Dkeycloak.migration.file=/tmp/keycloak-dump.json
```
4. import this realm when starting the Docker Container.

_or_

1. create a custom script as entry point which starts Keycloak in the background
2. poll 8080 using something like `nc -z $server 22 2>/dev/null;` until it's up.
3. create the users using `kcadm.sh`

I find both options to be a lot of effort for the simple task of adding test users for local development inside a Docker container. I'm kind of wondering how this seems to break only my workflow. Am I doing something fundamentally wrong?
